File: ECOO/ECOO '14 R3 P4 - Baby Talk.py


def checkBabyTalk(word):

    if len(word)%2 == 1:
        return False
    else:
        if word[:len(word)//2] == word[len(word)//2 :]:
            return True
        return False


for q in range(10):

    line = input()
    size = len(line)


    ans = 0

    dp = [0 for _ in range(size+2)]

    for i in range(1,size+1):
        for j in range(i):
            if checkBabyTalk(line[j:i]):
                dp[i] = max( dp[j] + len(line[j:i]), dp[i])

                if dp[i] > ans:
                    ans = dp[i]
                

    #print(dp)
    print(ans)
        

File: ECOO/ECOO '14 R3 P4 - Baby Talk.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){


    for ( int x = 0;x<10;x++){
        string word;
        cin>>word;
        int len = word.size();

        int dp[len+2];
        for(int q = 0;q<len+2;q++){
            dp[q] = 0;
        }


        int ans = 0;

        for (int i = 1;i<len+1;i++){


            for (int j = 0;j<i;j++){

                int newlen = i-j;
                string test = word.substr(j,newlen);


                if (newlen%2 == 0){
                    string first = test.substr(newlen/2);
                    string sec = test.substr(0,newlen/2);

                    //cout<<sec << " " << first <<" "<< j << " " << i << endl;

                    if (first == sec){
                        dp[i] = max(dp[i],dp[j] + newlen);

                        if (dp[i] > ans){
                            ans = dp[i];
                        }

                    }



                }




            }
        }

        cout<<ans<<endl;



    }


}
